collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala

```diff
@@ -150,7 +158,131 @@ object ImmutableArray extends StrictOptimizedSeqFactory[ImmutableArray] {
       ScalaRunTime.array_update(elements, i, f(i))
       i = i + 1
     }
-    new ImmutableArray(elements)
+    ImmutableArray.unsafeWrapArray(elements)
+  }
+
+  /** Wrap an existing `Array` into an `ImmutableArray` of the proper primitive specialization type without copying. */
+  def unsafeWrapArray[T](x: AnyRef): ImmutableArray[T] = (x match {
+    case null              => null
+    case x: Array[AnyRef]  => new ofRef[AnyRef](x)
+    case x: Array[Int]     => new ofInt(x)
+    case x: Array[Double]  => new ofDouble(x)
+    case x: Array[Long]    => new ofLong(x)
+    case x: Array[Float]   => new ofFloat(x)
+    case x: Array[Char]    => new ofChar(x)
+    case x: Array[Byte]    => new ofByte(x)
+    case x: Array[Short]   => new ofShort(x)
+    case x: Array[Boolean] => new ofBoolean(x)
+    case x: Array[Unit]    => new ofUnit(x)
+  }).asInstanceOf[ImmutableArray[T]]
```

Comment:
In the end we never statically know that we have an underlying `ofInt` instance. We always manipulate an `ImmutableArray[X]`. How does the JVM know that it doesn’t have to box/unbox elements?

Reply:
It doesn't. Any access through `ImmutableArray` is boxed. If you want unboxed access you need to cast down or instantiate a primitive subclass directly. It's the same for `WrappedArray`. In theory a `@specialized` annotation should be beneficial for both cases but since `WrappedArray` isn't currently specialized I didn't want to do any speculative changes without more complete benchmarks.